tests/test_TFNetworkLayer.py

```diff
@@ -150,6 +150,125 @@ def test_concat_sources_missing_dim():
     session.run(out.output.placeholder)
 
 
+def test_self_att_config():
+  with make_scope() as session:
+    n_in = 128
+    channels = 16
+    n_time = 128
+    n_batch = 10
+    context_size = 15
+    rnd = numpy.random.RandomState(42)
+    querry_dim = 32
+    querry_dim += context_size
+    key_dim = 32
+    value_dim = 32
+    num_heads = 4
+    key_scale = 1.0 / (float(key_dim) ** .5)
+
+    crop_left = (context_size - 1) // 2
+    crop_right = (context_size - 1) - crop_left
+
+    net_dict = {
+      "att_q": {"class": "linear", "activation": None, "with_bias": False, "n_out": querry_dim * num_heads, "from": "data:in0"},
+      "att_k": {"class": "linear", "activation": None, "with_bias": False, "n_out": key_dim * num_heads, "from": "data:in0"},
+      "att_v": {"class": "linear", "activation": None, "with_bias": False, "n_out": value_dim * num_heads, "from": "data:in0"},
+
+      "att_q_split": {"class": "split_dims", "axis": "f", "dims": (num_heads, querry_dim), "from": "att_q"},
+      "att_k_split": {"class": "split_dims", "axis": "f", "dims": (num_heads, key_dim), "from": "att_k"},
+      "att_v_split": {"class": "split_dims", "axis": "f", "dims": (num_heads, value_dim), "from": "att_v"},
+
+      "att_queries_crop": {"class": "slice", "axis": "t", "slice_start": crop_left, "slice_end": -crop_right, "from": "att_q_split"},
+
+      "att_queries_key_part": {"class": "slice", "axis": "f", "slice_start": 0, "slice_end": querry_dim - context_size, "from": "att_queries_crop"},
+      "att_queries_context_part": {"class": "slice", "axis": "f", "slice_start": querry_dim - context_size, "slice_end": querry_dim, "from": "att_queries_crop"},
+
+
+      "att_key_window": {"class": "window", "window_size": context_size, "padding": "valid", "from": "att_k_split"},
+      "att_energy_dot": {"class": "dot", "red1": "f", "red2": "f", "var1": None, "var2": "s:1", "from": ["att_queries_key_part", "att_key_window"]},
+      "att_energy_scaled": {"class": "eval", "eval": "source(0) * %f" % key_scale, "from": "att_energy_dot"},
+
+
+      "att_energy_biased": {"class": "combine", "kind": "add", "from": ["att_energy_scaled", "att_queries_context_part"]},
```

Comment:
Here, two inputs `att_energy_scaled`, `att_queries_context_part` have modified time dims. `att_energy_scaled` was windowed and then dot product was applied, `att_queries_context_part` was just sliced. The inputs can't be combined since they have different tags in time dim.

Reply:
Please show what they have (just show the output where we can see that).

Reply:
Exception creating layer <network via test_self_att_config>/'att_energy_biased' of class   CombineLayer with opts:
{'kind': 'add',
 'name': 'att_energy_biased',
 'network': <TFNetwork '<network via test_self_att_config>' train>,
 'output': Data(name='att_energy_biased_output', shape=(None, 8, 4, None, 15), batch_shape_meta=[B,T|u"time:var:u'att_key_window/Maximum:0':att_key_window/att_energy_scaled_output",8,4,'slice0:att_queries_crop',F|15]),
 'sources': [<EvalLayer 'att_energy_scaled' out_type=Data(shape=(None, 8, 4, 15), batch_shape_meta=[B,T|u"time:var:u'att_key_window/Maximum:0':att_key_window/att_energy_scaled_output",8,4,F|15])>,
             <SliceLayer 'att_queries_context_part' out_type=Data(shape=(None, 8, 4, 15), batch_shape_meta=[B,T|'slice0:att_queries_crop',8,4,F|15])>]}

Reply:
Also, you can see from the above error, that the output template is wrong, extra `None` dim is pasted.

Reply:
No, I just see that it is exactly correct, as intended. The inputs have two different time dims (`T|u"time:var:u'att_key_window/Maximum:0':att_key_window/att_energy_scaled_output"` vs `T|'slice0:att_queries_crop'`), thus, as already explained multiple times now, the common data shape will have both time dims.

Reply:
So, you say that everything works correct there. Ok, then I'll assume the tagging system as a black box. The question is, how to make my config work? It worked before, but now it doesn't.

Reply:
> So, you say that everything works correct there.

No. Please read carefully what I write. I just wrote that the output template is correct, exactly as intended. I did not say anything else. I just explained again why the output template is correct here.

> Ok, then I'll assume the tagging system as a black box.

No, don't do that. Why? It's not a black box. Read the code and please understand it (ask questions if sth is unclear). It's really simple, though. I already explained it multiple times now.

* If you have two different time dims, you also end up with two different time dims.
* If you have the same time dim, you end up with a single (the same) time dim.

That behavior is very simple, and that is exactly as it should be, and that works correct.
If the outcome of this is not what you want (having two different time dims), then by logical deduction, what you really want is that the time dim is not different but the same (for `att_energy_scaled` and `att_queries_context_part`). That is totally a different topic now. Why do you keep insisting that the logic of `get_common_data` is wrong? It's not. It's correct. The issue is somewhere else. Please try to understand that.

Thus, the real question is, why do `att_energy_scaled` and `att_queries_context_part` have different time dims?
And the next question is, how do you fix it (your config, or sth in Returnn) such that they have the same time dim?

> The question is, how to make my config work? It worked before, but now it doesn't.

Please just read what I wrote, and understand the real issue.

Reply:
> Why do you keep insisting that the logic of `get_common_data` is wrong? It's not. It's correct. The issue is somewhere else. Please try to understand that.
 
I agree, that `get_common_data` works fine. That's why I assume it as a black box. Then I do the tests with config, layer and imitate the behavior by means of Data.py.

Reply:
> I agree, that `get_common_data` works fine. That's why I assume it as a black box. Then I do the tests with config, layer and imitate the behavior by means of Data.py.

I don't understand. Why is it a black box? It's not. Just read the code. Also, I think it's actually pretty simple, what it does, and how it works.
Also, if you finally understand now that `get_common_data` (and also `CombineLayer`) work just fine, then maybe you can start searching where the real problem is.